ForgeUnity/Assets/BeardedManStudios/Editor/ForgeClassFieldValue.cs

```diff
@@ -220,6 +232,12 @@ public static ForgeAcceptableFieldTypes GetTypeFromAcceptable(string val)
 					return ForgeAcceptableFieldTypes.VECTOR3;
 				case "vector4":
 					return ForgeAcceptableFieldTypes.VECTOR4;
+				case "float2":
```

Comment:
Shouldn't these fields also de marked in `IsInterpolatable` above as interpolatable, just like Unity's `Vector` types?

I'm not too familiar with the internals, but an implementation for these in `BMSExtensions` and `BeardedMath` may be required, e.g.:

```csharp
public static bool Near(this Float4 target, Float4 other, float distance)
{
    return
        Near(target.x, other.x, distance) &&
        Near(target.y, other.y, distance) &&
        Near(target.z, other.z, distance) &&
        Near(target.w, other.w, distance);
}
```

```csharp
public static Float4 Lerp(Float4 a0, Float4 a1, float t)
{
    return new Float4(
        Lerp(a0.x, a1.x, t),
        Lerp(a0.y, a1.y, t),
        Lerp(a0.z, a1.z, t),
        Lerp(a0.w, a1.w, t)
    );
}
```

Note also that there appears to be a bug (https://github.com/BeardedManStudios/ForgeNetworkingRemastered/issues/384) currently that causes interpolation of these new types to break.

Reply:
I fixed #384 in another pull request, this makes interpolation "work" as in "not break any updates entirely anymore", but not pretty, as there is no actual interpolation happening as is the case for e.g. `Vector3`. I took a quick stab at this and implementing interpolation seemed to be fairly straightforward.

@DrPotatoNet I can't update this pull request's branch, so could you implement this patch on top of the changes, if this is okay for the maintainers of Forge?

```diff
From 21a5f83ca3e275ea88d42e5e50c51edb04cc0d2b Mon Sep 17 00:00:00 2001
From: MrWatts-Anorak <73220431+MrWatts-Anorak@users.noreply.github.com>
Date: Mon, 30 Nov 2020 12:55:37 +0100
Subject: [PATCH] Fix interpolation not working for position and rotation

There were multiple bugs here: interpolation was simply not implemented
for the new FloatX types, and falling back to a generic version. This
generic version also didn't work [1].

See also https://github.com/BeardedManStudios/ForgeNetworkingRemastered/pull/310#discussion_r532443679

[1] https://github.com/BeardedManStudios/ForgeNetworkingRemastered/issues/384
---
 .../Editor/ForgeClassFieldValue.cs            | 12 ++++++
 .../Scripts/Networking/BMSExtensions.cs       | 15 +++++++
 .../Scripts/StandAlone/Objects/Float2.cs      | 20 +++++++++-
 .../Scripts/StandAlone/Objects/Float3.cs      | 22 +++++++++-
 .../Scripts/StandAlone/Objects/Float4.cs      | 24 ++++++++++-
 .../StandAlone/Objects/InterpolateFloat2.cs   | 40 +++++++++++++++++++
 .../Objects/InterpolateFloat2.cs.meta         | 11 +++++
 .../StandAlone/Objects/InterpolateFloat3.cs   | 19 +++++++++
 .../Objects/InterpolateFloat3.cs.meta         | 11 +++++
 .../StandAlone/Objects/InterpolateFloat4.cs   | 19 +++++++++
 .../Objects/InterpolateFloat4.cs.meta         | 11 +++++
 11 files changed, 201 insertions(+), 3 deletions(-)
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs.meta
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs.meta
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs
 create mode 100644 unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs.meta

diff --git a/unity-wade/Assets/BeardedManStudios/Editor/ForgeClassFieldValue.cs b/unity-wade/Assets/BeardedManStudios/Editor/ForgeClassFieldValue.cs
index b5230f0..7375ed8 100644
--- a/unity-wade/Assets/BeardedManStudios/Editor/ForgeClassFieldValue.cs
+++ b/unity-wade/Assets/BeardedManStudios/Editor/ForgeClassFieldValue.cs
@@ -170,6 +170,15 @@ namespace BeardedManStudios.Forge.Networking.UnityEditor
 				case ForgeAcceptableFieldTypes.QUATERNION:
 					returnValue = "InterpolateQuaternion";
 					break;
+				case ForgeAcceptableFieldTypes.FLOAT2:
+					returnValue = "InterpolateFloat2";
+					break;
+				case ForgeAcceptableFieldTypes.FLOAT3:
+					returnValue = "InterpolateFloat3";
+					break;
+				case ForgeAcceptableFieldTypes.FLOAT4:
+					returnValue = "InterpolateFloat4";
+					break;
 				default:
 					returnValue = "Interpolated<" + baseTypeString + ">";
 					break;
@@ -188,6 +197,9 @@ namespace BeardedManStudios.Forge.Networking.UnityEditor
 				case ForgeAcceptableFieldTypes.VECTOR2:
 				case ForgeAcceptableFieldTypes.VECTOR3:
 				case ForgeAcceptableFieldTypes.VECTOR4:
+				case ForgeAcceptableFieldTypes.FLOAT2:
+				case ForgeAcceptableFieldTypes.FLOAT3:
+				case ForgeAcceptableFieldTypes.FLOAT4:
 				case ForgeAcceptableFieldTypes.QUATERNION:
 					returnValue = true;
 					break;
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/Networking/BMSExtensions.cs b/unity-wade/Assets/BeardedManStudios/Scripts/Networking/BMSExtensions.cs
index efc7b74..560e818 100644
--- a/unity-wade/Assets/BeardedManStudios/Scripts/Networking/BMSExtensions.cs
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/Networking/BMSExtensions.cs
@@ -57,6 +57,21 @@
 			return target.Between(other - distance, other + distance);
 		}

+		public static bool Near(this Float2 target, Float2 other, float threshold = 0.0012f)
+		{
+			return Float2.Distance(target, other) <= threshold;
+		}
+
+		public static bool Near(this Float3 target, Float3 other, float threshold = 0.0012f)
+		{
+			return Float3.Distance(target, other) <= threshold;
+		}
+
+		public static bool Near(this Float4 target, Float4 other, float threshold = 0.0012f)
+		{
+			return Float4.Distance(target, other) <= threshold;
+		}
+
 		public static bool Near<T>(this T target, T other, float distance)
 		{
 			return target.Equals(other);
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float2.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float2.cs
index 36ca8c2..bd04eeb 100644
--- a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float2.cs
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float2.cs
@@ -1,4 +1,6 @@
-﻿namespace BeardedManStudios
+﻿using System;
+
+namespace BeardedManStudios
 {
 	public partial struct Float2
 	{
@@ -33,5 +35,21 @@
 		{
 			return base.GetHashCode();
 		}
+
+		public static Float2 Lerp(Float2 a0, Float2 a1, float t)
+		{
+			return new Float2(
+				BeardedMath.Lerp(a0.x, a1.x, t),
+				BeardedMath.Lerp(a0.y, a1.y, t)
+			);
+		}
+
+		public static float Distance(Float2 a, Float2 b)
+		{
+			return (float)Math.Sqrt((double)(
+				Math.Pow(a.x - b.x, 2) +
+				Math.Pow(a.y - b.y, 2)
+			));
+		}
 	}
 }
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float3.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float3.cs
index 315f7b6..1670ab2 100644
--- a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float3.cs
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float3.cs
@@ -1,4 +1,6 @@
-﻿namespace BeardedManStudios
+﻿using System;
+
+namespace BeardedManStudios
 {
 	public partial struct Float3
 	{
@@ -34,5 +36,23 @@
 		{
 			return base.GetHashCode();
 		}
+
+		public static Float3 Lerp(Float3 a0, Float3 a1, float t)
+		{
+			return new Float3(
+				BeardedMath.Lerp(a0.x, a1.x, t),
+				BeardedMath.Lerp(a0.y, a1.y, t),
+				BeardedMath.Lerp(a0.z, a1.z, t)
+			);
+		}
+
+		public static float Distance(Float3 a, Float3 b)
+		{
+			return (float)Math.Sqrt((double)(
+				Math.Pow(a.x - b.x, 2) +
+				Math.Pow(a.y - b.y, 2) +
+				Math.Pow(a.z - b.z, 2)
+			));
+		}
 	}
 }
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float4.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float4.cs
index 2fe927e..d9bb279 100644
--- a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float4.cs
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/Float4.cs
@@ -1,4 +1,6 @@
-﻿namespace BeardedManStudios
+﻿using System;
+
+namespace BeardedManStudios
 {
 	public partial struct Float4
 	{
@@ -35,5 +37,25 @@
 		{
 			return base.GetHashCode();
 		}
+
+		public static Float4 Lerp(Float4 a0, Float4 a1, float t)
+		{
+			return new Float4(
+				BeardedMath.Lerp(a0.x, a1.x, t),
+				BeardedMath.Lerp(a0.y, a1.y, t),
+				BeardedMath.Lerp(a0.z, a1.z, t),
+				BeardedMath.Lerp(a0.w, a1.w, t)
+			);
+		}
+
+		public static float Distance(Float4 a, Float4 b)
+		{
+			return (float)Math.Sqrt((double)(
+				Math.Pow(a.x - b.x, 2) +
+				Math.Pow(a.y - b.y, 2) +
+				Math.Pow(a.z - b.z, 2) +
+				Math.Pow(a.w - b.w, 2)
+			));
+		}
 	}
 }
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs
new file mode 100644
index 0000000..5155500
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs
@@ -0,0 +1,40 @@
+﻿/*-----------------------------+------------------------------\
+|                                                             |
+|                        !!!NOTICE!!!                         |
+|                                                             |
+|  These libraries are under heavy development so they are    |
+|  subject to make many changes as development continues.     |
+|  For this reason, the libraries may not be well commented.  |
+|  THANK YOU for supporting forge with all your feedback      |
+|  suggestions, bug reports and comments!                     |
+|                                                             |
+|                               - The Forge Team              |
+|                                 Bearded Man Studios, Inc.   |
+|                                                             |
+|  This source code, project files, and associated files are  |
+|  copyrighted by Bearded Man Studios, Inc. (2012-2015) and   |
+|  may not be redistributed without written permission.       |
+|                                                             |
+\------------------------------+-----------------------------*/
+
+
+
+namespace BeardedManStudios.Forge.Networking.Unity
+{
+	public struct InterpolateFloat2 : IInterpolator<Float2>
+	{
+		public Float2 current;
+		public Float2 target;
+		public float LerpT { get; set; }
+		public bool Enabled { get; set; }
+		public ulong Timestep { get; set; }
+
+		public Float2 Interpolate()
+		{
+			if (!Enabled) return target;
+
+			current = Float2.Lerp(current, target, LerpT);
+			return current;
+		}
+	}
+}
\ No newline at end of file
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs.meta b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs.meta
new file mode 100644
index 0000000..f17feab
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat2.cs.meta
@@ -0,0 +1,11 @@
+fileFormatVersion: 2
+guid: 5d8f29f9be70c3a4e9c81a969eca280c
+MonoImporter:
+  externalObjects: {}
+  serializedVersion: 2
+  defaultReferences: []
+  executionOrder: 0
+  icon: {instanceID: 0}
+  userData:
+  assetBundleName:
+  assetBundleVariant:
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs
new file mode 100644
index 0000000..6599612
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs
@@ -0,0 +1,19 @@
+﻿namespace BeardedManStudios.Forge.Networking.Unity
+{
+	public struct InterpolateFloat3 : IInterpolator<Float3>
+	{
+		public Float3 current;
+		public Float3 target;
+		public float LerpT { get; set; }
+		public bool Enabled { get; set; }
+		public ulong Timestep { get; set; }
+
+		public Float3 Interpolate()
+		{
+			if (!Enabled) return target;
+
+			current = Float3.Lerp(current, target, LerpT);
+			return current;
+		}
+	}
+}
\ No newline at end of file
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs.meta b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs.meta
new file mode 100644
index 0000000..0ad8ee4
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat3.cs.meta
@@ -0,0 +1,11 @@
+fileFormatVersion: 2
+guid: 57d00189c59006d48a08c6c9ed915b29
+MonoImporter:
+  externalObjects: {}
+  serializedVersion: 2
+  defaultReferences: []
+  executionOrder: 0
+  icon: {instanceID: 0}
+  userData:
+  assetBundleName:
+  assetBundleVariant:
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs
new file mode 100644
index 0000000..28622e0
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs
@@ -0,0 +1,19 @@
+﻿namespace BeardedManStudios.Forge.Networking.Unity
+{
+	public struct InterpolateFloat4 : IInterpolator<Float4>
+	{
+		public Float4 current;
+		public Float4 target;
+		public float LerpT { get; set; }
+		public bool Enabled { get; set; }
+		public ulong Timestep { get; set; }
+
+		public Float4 Interpolate()
+		{
+			if (!Enabled) return target;
+
+			current = Float4.Lerp(current, target, LerpT);
+			return current;
+		}
+	}
+}
\ No newline at end of file
diff --git a/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs.meta b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs.meta
new file mode 100644
index 0000000..d6b9aee
--- /dev/null
+++ b/unity-wade/Assets/BeardedManStudios/Scripts/StandAlone/Objects/InterpolateFloat4.cs.meta
@@ -0,0 +1,11 @@
+fileFormatVersion: 2
+guid: 77b17d7f89db5bb4bb8038dfaa3b6da9
+MonoImporter:
+  externalObjects: {}
+  serializedVersion: 2
+  defaultReferences: []
+  executionOrder: 0
+  icon: {instanceID: 0}
+  userData:
+  assetBundleName:
+  assetBundleVariant:
--
2.28.0.windows.1
```